### DISCUSSION TITLE
Backend API updates

### DISCUSSION THREAD

**Discussion by ocebenzer** [bounswe/bounswe2020group9#212] [General]
As backend team, we plan to announce the critical changes we do on the API system.
As these updates can break both frontend and mobile, we hope this issue will help to handle possible problems.

As time goes on, each update's header will be edited whether they are live or not.

**Comment by ocebenzer** [comment]
##  Live:
Product JSON now has parameter: "detail", this holds a description for that product:
```
    {
        "id": 3,
        "name": "Iphone XS 64GB",
        "brand": "Apple",
        "labels": [
            "10% off"
        ],
        "detail": "The new Iphone, 64gb version. It is quite beautiful",
        "category": "default",
        "price": 11211.52,
        "stock": 48,
        "sell_counter": 0,
        "rating": 0.0,
        "vendor": 2,
        "picture": null
    }
```
note that it is not needed to be filled, it can be blank as such:
```
...
"detail": "",
...
```

**Comment by ocebenzer** [comment]
##  Live:
Category Update! now each product can hold only one category, but that category has a string named `parent`:
```
{
        "id": 3,
        "name": "Iphone XS 64GB",
        "brand": "Apple",
        "labels": [
            "10% off"
        ],
        "detail": "The new Iphone, 64gb version. It is quite beautiful",
        "category":{
            "name":"Books",
            "parent":"Living"
                }
        }
        "price": 11211.52,
        "stock": 48,
        "sell_counter": 0,
        "rating": 0.0,
        "vendor": 2,
        "picture": null
    }
```

A category without a parent will be defined as an empty string: `"parent":""`

**Reply by ocebenzer** [reply]
## Announcement
category object now also returns "id":
```
    "category": {
        "name": "Sports",
        "parent": "Living",
        "id": 35
    },
```

so the final JSON above will be as such:
```
{
        "id": 3,
        "name": "Iphone XS 64GB",
        "brand": "Apple",
        "labels": [
            "10% off"
        ],
        "detail": "The new Iphone, 64gb version. It is quite beautiful",
        "category": {
            "name": "Sports",
            "parent": "Living",
            "id": 35
        },
        "price": 11211.52,
        "stock": 48,
        "sell_counter": 0,
        "rating": 0.0,
        "vendor": 2,
        "picture": null
    }
```
This was needed since POST product request requires `category_id`.

**Comment by ocebenzer** [comment]
## Live, needs validation:
PUT request on /api/user/profile/ is not working as intended:
* first_name last_name, not mandatory but can be changed
* password check, if true, reset password to the new_password

**Comment by ocebenzer** [comment]
##  Live:
### POST Product:
POST /api/product/ is working as intended, picture is set to null by default

must be logged in as a vendor in order to POST Product

returns 201 if created succesfully

returns 400 if failed

**Comment by ocebenzer** [comment]
## Live
Right now, you can't add category to the product, it defaults into "Categories".

We plan to fix this  to such:
```
{
POST {{url}}/api/product/
HEADER "Authorization":"Token ..."
BODY:
{
    "name":"myproduct",
    "brand":"mybrand",
    "price":3.95,
    "stock":10000,
    "category_id":5
}
}
```
note the added `category_id` parameter

This also implies that we need to add a GET _all list names_ API call

This might change in the future, it will be reported if it does.

**Comment by ocebenzer** [comment]
## Live:

Cart system avalable!

now /api/user/cart/ can handle GET, POST, PUT, DELETE requests as intended

please look them up from the [API Documentation](https://github.com/bounswe/bounswe2020group9/wiki/API-Documentation)

**Comment by ibrahimorhan** [comment]
Search system is available!
url and all descriptions are in [API Documentation](https://github.com/bounswe/bounswe2020group9/wiki/API-Documentation) with example usages

**Comment by ibrahimorhan** [comment]
Google user sign in and sign up is available

**Comment by ibrahimorhan** [comment]
Reset password system has changed.  for "Forgot Password" functionality "/resetpwmail" and "/resetpw" will be used. for resetting password from profile page "/resetpwprofile" will be used. 
Descriptions are in API Documentation

**Comment by ibrahimorhan** [comment]
/googleuser API now requires also token given by Google

**Comment by mehmetkemaloglu** [comment]
Messaging and notification systems are added to API documentation. The pull request didn't merged and the updates didn't deployed yet. Therefore, keep in the mind that the API documentation is subject to change.

**Comment by mehmetkemaloglu** [comment]
New messaging system API is added to API documentation. 
The API 'GET's all message data of the user.

**Comment by mehmetkemaloglu** [comment]
Messaging system API is updated. Timestamp orders are resolved. 
New feature: GET all messages API additionally returns number of new messages of each conversation.

**Comment by mehmetkemaloglu** [comment]
GET vendor location API is implemented. You can check it from API documentation. Updates will be available after the deployment.

**Comment by mehmetkemaloglu** [comment]
GET all comments of a user API is available at backend master. You can check it from API documentation.

**Comment by mehmetkemaloglu** [comment]
Last API is updated.
